Alerting: Migration UI

soniaAguilarPeiron opened the pull request:
**What is this feature?**

This PR adds a migration tool to migrate datasource-managed rules to GMA.

The feature is behind a feature toggle `alertingMigrationUI`

You need to enable `alertingConversionAPI` ff for being able to use the new api endpoint.

The migration UI is only available for users with `alert.rules:create` and `alert.rules.external:read` permissions.
We only allow to migrate from `datasources with ruler`, because the conversion endpoint uses the ruler API for the migration.

The required parameters for the conversion are: the data source and the target folder. 
Additionally, user can filter by namespace and group.
User can also decide if the new GMAlerts will be paused once they are converted.

**Why do we need this feature?**

This is a required tool in order to move forward to [Single Alert Type](https://github.com/grafana/alerting-squad/issues/971).

**Who is this feature for?**

Alerting users that want to migrate DS-managed rules to GMA.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/alerting-squad/issues/1033

**Special notes for your reviewer:**


https://github.com/user-attachments/assets/569b37fd-5b22-4ef6-968e-e35dee990209


Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

soniaAguilarPeiron committed:
add feature toggle alertingMigrationUI

soniaAguilarPeiron committed:
wip

soniaAguilarPeiron committed:
wip

soniaAguilarPeiron committed:
wip2

soniaAguilarPeiron committed:
add button to export in alert list and add target folder in the form

soniaAguilarPeiron committed:
update namings

soniaAguilarPeiron committed:
protext migration route with alertingMigrationUI feature toggle

soniaAguilarPeiron committed:
remove unused error style

soniaAguilarPeiron committed:
add namspace and group filter

soniaAguilarPeiron committed:
fix translations and prettier

soniaAguilarPeiron committed:
fix duplication

soniaAguilarPeiron committed:
use the new endpoint

soniaAguilarPeiron committed:
update right roles for the FE route

soniaAguilarPeiron committed:
refactor

soniaAguilarPeiron committed:
notify success

gillesdemey reviewed (approved):
UI changes LGTM :shipit:

gillesdemey reviewed (changes requested):
Pulled the trigger on the review a bit early; so setting this to "request changes" for now

gillesdemey commented:
Couple of comments on the import flow (also discussed offline);

1. we should look into allowing to import to the root folder and setting that as the default
2. make it more obvious that choosing a namespace / group is optional and will result in a partial import

other items to be determined;
----

1. show a banner somewhere indicating that we are going to move forward with GMA rules and asking the user if they wish to migrate their data-source managed alert rules
2. Allow choosing what folder to "map" a data source namespace to so the user doesn't have to go through the import UI several times to achieve this
3. (technical) investigate adding the ability to import from vanilla Prometheus – this needs some more discussion because we can't disabled / remove the alert rules from the source.

soniaAguilarPeiron committed:
address feedback

soniaAguilarPeiron committed:
fix translations

gillesdemey committed:
Simplify migration UI for alert rules to GMA format

Extracts migration button component and reuses it in banner for better
code organization and consistency.

gillesdemey committed:
Fix root folder check in import alert rules UI

Replace hardcoded empty string check with lodash isEmpty for folder UID
validation

gillesdemey committed:
Add collapsible section for import rule options

soniaAguilarPeiron commented on public/app/features/alerting/routes.tsx:
```diff
@@ -209,6 +211,18 @@ export function getAlertingRoutes(cfg = config): RouteDescriptor[] {
+      path: '/alerting/import-datasource-managed-rules',
+      roles: evaluateAccess([AccessControlAction.AlertingRuleCreate, AccessControlAction.AlertingRuleExternalRead]),
+      component: config.featureToggles.alertingMigrationUI
+        ? SafeDynamicImport(
```
opssss

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/import-to-gma/ImportFromDSRules.tsx:
```diff
@@ -0,0 +1,210 @@
+
+const getStyles = (theme: GrafanaTheme2) => ({
+  buttonSpinner: css({
+    marginRight: theme.spacing(1),
```
💯

soniaAguilarPeiron committed:
address feedback

soniaAguilarPeiron committed:
Split useAlertRuleSuggestions, one for labels, and other for namespaces

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/import-to-gma/NamespaceAndGroupFilter.tsx:
```diff
@@ -0,0 +1,163 @@
+
+  const style = useStyles2(getStyle);
+  const namespace = watch('namespace');
+  const { namespaceGroups, isLoading } = useGetNameSpacesByDatasourceName(rulesSourceName);
```
[commit](https://github.com/grafana/grafana/pull/102010/commits/8f1597dcabf54a652058fe68221addb52acc249b) pushed!

gillesdemey reviewed (dismissed):
Front-end changes LGTM!

soniaAguilarPeiron committed:
refactor splited hooks

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/import-to-gma/ImportFromDSRules.tsx:
```diff
@@ -0,0 +1,210 @@
+        targetFolderUID: data.targetFolder!.uid,
+        pauseRecordingRules: data.pauseRecordingRules,
+        pauseAlerts: data.pauseAlertingRules,
+        ...(data.namespace ? { namespace: data.namespace } : {}),
```
I think it's a common pattern in our Grafana code, and it ensures that only defined values are included in the object.
If we pass undefined as a parameter instead of using this pattern, the namespace property would be included in the object with undefined value.This might cause issues depending on how the receiving function or API handles undefined values.

tomratcliffe commented on public/app/features/alerting/unified/components/import-to-gma/ImportFromDSRules.tsx:
```diff
@@ -0,0 +1,205 @@
+                  <Controller
+                    render={({ field: { onChange, ref, ...field } }) => (
+                      <Stack width={50}>
+                        <NestedFolderPicker
```
The picker isn't able to overflow out of the collapsing section btw:
![image](https://github.com/user-attachments/assets/53b9208b-f19c-422e-9200-e2dbbbad9a61)

tomratcliffe replied:
(I have a fix for this, will do in a separate PR)

tomratcliffe replied:
https://github.com/grafana/grafana/pull/102634 for the fix for this

tomratcliffe commented on public/app/features/alerting/unified/components/rules/CloudRules.tsx:
```diff
@@ -48,24 +50,39 @@ export const CloudRules = ({ namespaces, expandAll }: Props) => {
+  const canViewCloudRules = viewExternalRuleSupported && viewExternalRuleAllowed;
+
+  const canCreateGrafanaRules = createRuleSupported && createRuleAllowed;
+  const canMigrateToGMA = canCreateGrafanaRules && canViewCloudRules && config.featureToggles.alertingMigrationUI;
```
Should this also check that there are datasources configured that the user can migrate? e.g. include `hasDataSourcesConfigured`

tomratcliffe commented on public/app/features/alerting/unified/components/import-to-gma/NamespaceAndGroupFilter.tsx:
```diff
@@ -0,0 +1,121 @@
+  rulesSourceName: string;
+}
+
+export const NamespaceAndGroupFilter = ({ rulesSourceName }: Props) => {
```
This is almost the same as `components/rule-editor/GroupAndNamespaceFields.tsx` - could we combine the two?

tomratcliffe replied:
Actually, looking at this more - we could probably move this to use `ComboBox` instead, as this one still lets you create a namespace/group by typing and pressing enter (which doesn't make sense in this context)

tomratcliffe commented on public/app/features/alerting/unified/components/import-to-gma/NamespaceAndGroupFilter.tsx:
```diff
@@ -0,0 +1,121 @@
+      <Field
+        data-testid="namespace-picker"
+        label={t('alerting.import-to-gma.namespace.label', 'Namespace')}
+        // Disable translations as we don't intend to use this dropdown longterm,
+        // so avoiding us adding translations for the sake of it
+        // eslint-disable-next-line @grafana/no-untranslated-strings
+        description={t('alerting.import-to-gma.namespace.description', 'Type to search for an existing namespace')}
```
We ended up translating these here - I guess we do expect this approach to remain for a little longer? The other component had the rule commented out because ideally we want to make the namespace/group selection look more like the GMA approach

soniaAguilarPeiron replied:
we have already translated..removing the comment

tomratcliffe committed:
Alerting: (PR review) Redesign slightly + use combobox (#102614)

Redesign slightly + use combobox

soniaAguilarPeiron committed:
udpate naming

soniaAguilarPeiron committed:
remove eslint-disable-next-line @grafana/no-untranslated-strings and add hasDataSourcesConfigured as a condition to render the banner

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/useAlertRuleSuggestions.tsx:
```diff
@@ -32,34 +32,61 @@ export function useAlertRuleSuggestions(rulesSourceName: string) {
+  const namespaceGroups = useMemo(() => {
     if (isPrometheusRulesLoading || isRulerRulesLoading) {
-      return new Map<string, Set<string>>();
+      return new Map<string, string[]>();
```
Is `string[]` better than `Set<string>`? `Set<string>` allow us to avoid duplicates

soniaAguilarPeiron replied:
ops, I think this change is not correct

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/useAlertRuleSuggestions.tsx:
```diff
@@ -32,34 +32,61 @@ export function useAlertRuleSuggestions(rulesSourceName: string) {
 }
 
-function promNamespacesToNamespaceGroups(promNamespaces: RuleNamespace[]) {
+export function promNamespacesToNamespaceGroups(promNamespaces: RuleNamespace[]) {
```
Any reason to export these functions? If we want to use them in other places then it might be worth to move them to a more appropriate file

soniaAguilarPeiron committed:
use button instead of banner

soniaAguilarPeiron committed:
fix routes

soniaAguilarPeiron committed:
use new api

soniaAguilarPeiron committed:
update api

soniaAguilarPeiron committed:
remove query params

soniaAguilarPeiron committed:
update translations

soniaAguilarPeiron committed:
update ff files

soniaAguilarPeiron committed:
skip subpath when redirecting to the alert list view with filters

alexander-akhmetov commented:
/deploy-to-hg

soniaAguilarPeiron committed:
prettier

soniaAguilarPeiron committed:
disable provenance

soniaAguilarPeiron commented:
/deploy-to-hg

alexander-akhmetov commented:
/deploy-to-hg

soniaAguilarPeiron committed:
prettier

soniaAguilarPeiron committed:
Don't send target forlder uid in the payload as undefined if it's not selected

soniaAguilarPeiron commented:
/deploy-to-hg

soniaAguilarPeiron committed:
prettier

soniaAguilarPeiron committed:
add confirmation modal with a preview

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into alerting/migration-to-gfm-rules

soniaAguilarPeiron commented:
/deploy-to-hg

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/import-to-gma/ImportFromDSRules.tsx:
```diff
@@ -0,0 +1,205 @@
+  };
+
+  const notifyApp = useAppNotification();
+  const [optionsShowing, toggleOptions] = useToggle(false);
```
this is outdated, the notification has been moved to the `ConfirmConversionModal`

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/import-to-gma/ImportFromDSRules.tsx:
```diff
@@ -0,0 +1,205 @@
+                    {...field}
+                    onChange={(ds: DataSourceInstanceSettings) => {
+                      setValue('selectedDatasourceUID', ds.uid);
+                      setValue('selectedDatasourceName', ds.name);
```
is not enough, we need it for getting namespaces for this datasource when using `useGetNameSpacesByDatasourceName`

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/rule-editor/useAlertRuleSuggestions.tsx:
```diff
@@ -32,34 +32,61 @@ export function useAlertRuleSuggestions(rulesSourceName: string) {
+  const [fetchRulerRules, { data: rulerRules = emptyRulerConfig, isLoading: isRulerRulesLoading }] =
+    useLazyRulerRulesQuery();
+
+  const { data: promNamespaces = [], isLoading: isPrometheusRulesLoading } = usePrometheusRuleNamespacesQuery(
+    { ruleSourceName: rulesSourceName },
+    { skip: !prometheusRulesPrimary }
+  );
+
+  useEffect(() => {
+    if (features?.rulerConfig && !prometheusRulesPrimary) {
+      fetchRulerRules({ rulerConfig: features.rulerConfig });
+    }
+  }, [features?.rulerConfig, fetchRulerRules]);
```
not sure what do you mean? this code was only moved, is an [existing code](https://github.com/grafana/grafana/blob/a65cc0df93c5568d0ba32f0171a3c2cd04997cf7/public/app/features/alerting/unified/components/rule-editor/useAlertRuleSuggestions.tsx#L24-L33)

soniaAguilarPeiron committed:
remove unnecessary export

tomratcliffe committed:
Add target datasource UID for recording rules

tomratcliffe committed:
Use skipToken

tomratcliffe committed:
Add recording rule datasource and update designs

tomratcliffe committed:
Use different hook to fetch rulerRules

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/migration-to-gfm-rules

tomratcliffe commented:
/deploy-to-hg

tomratcliffe reviewed (approved):
LGTM, I'll address additional points in the followup PR!
